test/cli/pin.js

```diff
@@ -0,0 +1,148 @@
+/* eslint-env mocha */
+'use strict'
+
+const expect = require('chai').expect
+const series = require('async/series')
+const runOnAndOff = require('../utils/on-and-off')
+const _ = require('lodash')
+
+// use a tree of ipfs objects for recursive tests:
+//  root
+//   |`leaf
+//    `branch
+//      `subLeaf
+
+const filenames = [ 'root', 'leaf', 'branch', 'subLeaf' ]
+
+const keys = {
+  root: 'QmWQwS2Xh1SFGMPzUVYQ52b7RC7fTfiaPHm3ZyTRZuHmer',
+  leaf: 'QmaZoTQ6wFe7EtvaePBUeXavfeRqCAq3RUMomFxBpZLrLA',
+  branch: 'QmNxjjP7dtx6pzxWGBRCrgmjX3JqKL7uF2Kjx7ExiZDbSB',
+  subLeaf: 'QmUzzznkyQL7FjjBztG3D1tTjBuxeArLceDZnuSowUggXL'
+}
+
+describe('pin', () => runOnAndOff((thing) => {
+  const filesDir = 'test/test-data/tree/'
+
+  let ipfs
+
+  before((done) => {
+    ipfs = thing.ipfs
+
+    ipfs(`object put ${filesDir + 'root.json'}`)
+      .then(() => ipfs(`object put ${filesDir + 'root.json'}`))
+      .then(() => ipfs(`object put ${filesDir + 'leaf.json'}`))
+      .then(() => ipfs(`object put ${filesDir + 'branch.json'}`))
+      .then(() => ipfs(`object put ${filesDir + 'subLeaf.json'}`))
+      .then(() => {
+        done()
+      })
+  })
+
+  it('add (recursively by default)', (done) => {
+    console.log('cli pin add starting')
+    ipfs(`pin add ${keys.root}`).then((out) => {
+      console.log('cli pin add callback')
+      console.log(out)
+//      expect(out).to.eql(`pinned ${keys.root} recursively\n`)
```

Comment:
For the core tests, they pass locally for me if I make the changes discussed [here](https://github.com/ipfs/interface-ipfs-core/pull/63).